Detect non-concretized broadcast domains

naoyam opened the pull request:
Detect non-concretized broadcast domains so that unnecessary parallel broadcast operations should not be generated.

Needed to clean up the thread predicate info passed to Kernel. The only necessary information is which parallel types are in fact parallel in broadcast ops, so a map holding that information is added to KernelSummary. ThreadPredicateMap is dropped from Kernel.

The repro of #1410 no longer generates a grid broadcast call.

Closes #1410 and #1411

Commit:
WIP

csarofeen commented on torch/csrc/jit/codegen/cuda/lower_trivial_broadcast.cpp:
```diff
@@ -0,0 +1,176 @@
+  ConcretizedBroadcastDomains(Fusion* fusion) {
+    for (const auto fusion_input_tv :
+         ir_utils::filterByType<TensorView>(fusion->inputs())) {
+      for (auto root_id : fusion_input_tv->getRootDomain()) {
```
Is this okay to not be rfactor, if a fusion input is the input to a view, does this track through merges? I imagine we could merge broadcast root domains with non-broadcast root domains. On second thought I think this would be normally tracked through replay tracking.

csarofeen replied:
Yeah, I think this is fine.

rdspring1 replied:
We shouldn't merge broadcast root domains with non-broadcast root domains. For that case, we remove them with a trivial reduction before the view op.

https://github.com/csarofeen/pytorch/blob/devel/torch/csrc/jit/codegen/cuda/ops/alias.cpp#L63-L65

Broadcast root domains could be passed through the view to the rfactor domain.

naoyam replied:
Added a test with view.

I have a question on merging broadcast and non-broadcast domains. Please see #1418

csarofeen commented on torch/csrc/jit/codegen/cuda/lower_trivial_broadcast.cpp:
```diff
@@ -0,0 +1,176 @@
+      trivial_bops_.insert(bop);
+      return;
+    }
+    bool is_trivial = true;
```
I assume you're not done yet, but I think the logic is check all inputs of the parallel broadcast dimensions. If none of them are concretized then it's a trivial broadcast op.

naoyam replied:
Yes, the code below should do that. I'll clean it up and add comments.

Commit:
Detect non-concretized broadcast domains

Used to avoid unnecessary parallel broadcast operations.

Needed to clean up the thread predicate info passed to Kernel. The only
necessary information is which parallel types are in fact parallel in
broadcast ops, so a map holding that information is added to
KernelSummary. ThreadPredicateMap is dropped from Kernel.

naoyam commented:
Still working on this. 

I mentioned I was planning to replace trivial broadcasts with Set. I realized it's not a good idea. Broadcast adds a new root domain, whereas Set does not. The fact that an expression is `BroadcastOp` is used to detect new broadcast domains (along with `BroadcastOp::getBroadcastDimFlags()`. Such information would be lost if translated to `SetOp`.

naoyam commented on torch/csrc/jit/codegen/cuda/ir_iostream.cpp:
```diff
@@ -521,6 +521,7 @@ void IrPrinter::handle(const kir::TensorIndex* ti) {
     }
   }
   os_ << "]";
+  os_ << " view( T" << varName(ti->view()) << " )";
```
I think it's useful to show which TensorView this TensorIndex corresponds to.

csarofeen replied:
Would it be better to make this a little simpler as: `"T"<<ti->view()->name()`

csarofeen replied:
I'm blind, you had it right 🤦‍♂️

naoyam commented on torch/csrc/jit/codegen/cuda/kernel.cpp:
```diff
@@ -263,8 +273,6 @@ class ValidateAllocation : private OptOutConstDispatch {
 void Kernel::finalize(std::vector<Expr*> top_level_exprs) {
   TORCH_INTERNAL_ASSERT(top_level_exprs_.empty());
   top_level_exprs_ = std::move(top_level_exprs);
-  predicate_map_ = std::make_unique<ThreadPredicateMap>(
```
No longer necessary.

naoyam commented on torch/csrc/jit/codegen/cuda/kernel.h:
```diff
@@ -72,6 +74,10 @@ struct KernelSummary {
   std::vector<std::pair<const Val*, const Val*>> splits_to_validate;
+
+  //! Effective ParallelTypes of broadcast ops
+  std::unordered_map<const BroadcastOp*, ParallelTypeBitmap>
```
This effectively replaces `Kernel::predicate_map_`.

naoyam commented on torch/csrc/jit/codegen/cuda/lower2device.h:
```diff
@@ -132,12 +142,14 @@ class TORCH_CUDA_CU_API GpuLower : public NonCopyable {
   ComputeAtMap ca_index_map_;
   ComputeAtMap ca_parallel_map_;
   TrivialReductionInfo trivial_reduction_info_;
+  TrivialBroadcastInfo trivial_broadcast_info_;
```
Turns out this isn't used. Will remove.

naoyam commented on torch/csrc/jit/codegen/cuda/lower_trivial_broadcast.h:
```diff
@@ -0,0 +1,77 @@
+ private:
+  //! Set of trivial broadcast domains
+  std::unordered_set<BroadcastOp*> trivial_bops_;
+};
```
It turned out that this isn't needed. Is there any potential use case? Otherwise, I'll remove it.

csarofeen replied:
I think doing what you do in codegen is more convenient than blanket marking broadcast ops as trivial or not. The only thing I might suggest in a follow up is just avoid generating GridBroadcast is it's not broadcasted on the grid. I wonder if we should have explicit blockBroadcast/Reduce nodes, so we can do the same and avoid the complexity later in codegen. If so to be done in a follow up PR, not this one.

csarofeen replied:
Removed this code.

csarofeen replied:
Naoya mentioned to me `lower_index` checks `getParallelBroadcastDomains` before it generates a grid broadcast, so it won't generate a node anymore if not a real grid broadcast.

naoyam commented on test/cpp/jit/test_gpu.cpp:
```diff
@@ -20268,6 +20268,173 @@ TEST_F(NVFuserTest, FusionIntermediateTensorVectorize_CUDA) {
+
+  fusion.printMath();
+  fusion.printKernel();
+}
```
Wanted to test merged broadcast domains, but this test fails even with `devel`. See #1418

csarofeen commented on torch/csrc/jit/codegen/cuda/lower_trivial_broadcast.cpp:
```diff
@@ -0,0 +1,176 @@
+  // Propagate broadcast origin info from producers to consumers
+  for (auto producer : ir_utils::filterByType<TensorView>(expr->inputs())) {
+    std::unordered_set<IterDomain*> producer_broadcasts;
+    for (auto producer_id : producer->getMaybeRFactorDomain()) {
```
Can you mention this assumes broadcast axes are not merged ~with non-broadcast axes~ between root and rfactor?

csarofeen replied:
Added a comment.

csarofeen commented on torch/csrc/jit/codegen/cuda/lower_trivial_broadcast.cpp:
```diff
@@ -0,0 +1,176 @@
+    auto child = child_domains.front();
+    child_domains.pop_front();
+    const auto& child_uses = child->uses();
+    if (child_uses.size() == 1 && !child_uses[0]->outputs().empty() &&
```
Why not just go through all uses and mark every iter domain as concretized?
Couldn't we even use `DependencyChain::getAllDependentVals` to do this?

csarofeen replied:
`DependencyChain::getAllDependentVals` goes in the wrong direction, and the use versions in dependency check just do fusion traversal which starts at terminating outputs, so we'd have to enable the full recursive version of iter visitor which I'm not really ready to do because of perf impact.

csarofeen replied:
I made this more aggressive in the last commit I pushed, if you don't like this or think it's wrong feel free to revert. I didn't think this condition was important, but I might be wrong as I didn't understand why.

csarofeen commented on torch/csrc/jit/codegen/cuda/lower_trivial_broadcast.cpp:
```diff
@@ -0,0 +1,176 @@
+    auto out_tv = bop->out()->as<TensorView>();
+    // ConcretizedBroadcastDomains is used inside
+    // getParallelBroadcastDomains so non-concrete domains are not returned.
+    const auto parallel_types = pred_map.getParallelBroadcastDomains(out_tv);
```
nit on the name: `parallel_broadcast_ids`

csarofeen replied:
Code was removed based on https://github.com/csarofeen/pytorch/pull/1412#discussion_r794234903

csarofeen reviewed (commented):
Minor changes requested.
Still thinking about the merging of bcast-iter domain issue

csarofeen committed:
Disable test for now, mark as todo and link to issue.

csarofeen committed:
Minor cleanup.

csarofeen committed:
Be more aggressive on marking IDs as concretized bcast.

csarofeen committed:
format

naoyam commented on torch/csrc/jit/codegen/cuda/lower_trivial_broadcast.cpp:
```diff
@@ -96,27 +96,27 @@ void ConcretizedBroadcastDomains::handle(Expr* expr) {
+    for (auto child_use : child_uses) {
+      for (auto out_id :
+           ir_utils::filterByType<IterDomain>(child_use->outputs())) {
+        if (visited.find(out_id) == visited.end()) {
```
It seems `visited` never gets inserted. This seems to continue traversing as long as `IterDomain` is found in the outputs of `child_use`.

csarofeen committed:
Minor cleanup.
